Added Buildkite pipeline and integration tests (#2)
rust-vmm/rust-vmm-ci: README.md

# rust-vmm-ci

> The `rust-vmm-ci` contains integration tests and [Buildkite]() pipeline

Reviewer: The link is missing.

definitions that are used for running the CI for all rust-vmm crates.

> Having a centralized place for the tests is one of the enabler for keeping the

Reviewer: ```suggestion
Having a centralized place for the tests is an enabler for keeping the
```

same quality standard for all crates in rust-vmm.

## Getting Started with rust-vmm-ci

To run the integration tests defined in the pipeline as part of the CI:

1. Add rust-vmm-ci as a git submodule to your repository

```bash
git submodule add https://github/rust-vmm/rust-vmm-ci
git add rust-vmm-ci/
git commit -s -m "Added rust-vmm-ci as submodule"
```

> 

Reviewer 2: You may want to briefly what that does and why we do this.

Author: Do you mean the submodule?

Reviewer 2: I mean to explain what does it actually do and why we're doing it. A readable explanation after the set of commands.
But that's a nit.

  separated by one coma. If the repository does not have any autogenerated
  files, *exclude_path* should be an empty string.
- *crate_features*: `cargo kcov` does not build crate features by default. To
  get the coverage report including optional features these need to be
  specified in *crate_features* separated by coma. If the crate does not have
  any features, this field should be empty.

This file is required for the coverage integration so it needs to be added
to the repository as well.

3. Create a new pipeline definition in Buildkite. For this step ask one of the
rust-vmm Buildkite [admins](CODEOWNERS) to create one for you. Add a pipeline
step that is uploading the rust-vmm-ci pipeline:

```bash
buildkite-agent pipeline upload rust-vmm-ci/.buildkite/pipeline.yml
```

4. The code owner of the repository will have to setup a WebHook for triggering
the CI on Pull Requests and pushes.

## Buildkite Pipeline

The Buildkite pipeline is the definition of integration tests to be run as part
of the CI. It includes steps for running unit tests and linters (including
coding style checks), and computing the coverage.

> 

Reviewer 2: This section is a little confusing to me. It feels like there's a need to define that pipeline whereas in reality this is what the git submodule brings.
So I would refer back to step 1 of the previous section, explaining that this steps adds a buildkite pipeline into the repo and that buildkite will use it.

Author: This section is not about using rust-vmm-ci, but about how we created rust-vmm-ci and what does it have. It's a "developer" readme. Should I put it in a different .md file and keep the readme only with information about how to use rust-vmm-ci?

Author: I did some rephrasing here and there. After some more thinking I decided it's better to have only one readme for now. Maybe we should have both sections (Buildkite Pipeline and Integration Tests) in another section, but I can't find a good name. Maybe "rust-vmm-ci development"? What do you say?

Reviewer 2: +1 for one single readme.
This looks fine to me.

Example of step that checks the build:

```yaml
steps:
  - label: "build-gnu-x86"
    commands:
     - cargo build --release
    retry:
      automatic: false
    agents:
      platform: x86_64.metal
    plugins:
      - docker#v3.0.1:
          image: "rustvmm/dev:v2"
          always-pull: true
```

To see all steps in the pipeline check the
[.buikite/pipeline.yml](.buildkite/pipeline.yml) file.

### Custom Repository Hooks

> 

Reviewer 2: We should also define a section for adding a custom buildkite pipeline. The linux-loader (again) for example needs a custom one in order to build all features combinations.

Author: Added section: https://github.com/rust-vmm/rust-vmm-ci/blob/5d90acfbcb4166cb1e5dbd2487108a3e9aea0f79/README.md#custom-pipeline

Reviewer 2: Thanks.

The integration tests of some repositories have dependencies on external
resources. One example is
[linux-loader](https://github.com/rust-vmm/linux-loader/) which needs to
download a bzImage before running the unit tests. Because this is specific
to the linux-loader crate, the logic for downloading the required resources
cannot be part of the common pipeline. The mechanism used here is
[Repository Hooks](https://buildkite.com/docs/agent/v3/hooks#repository-hooks).
The hooks are defined per repository and live in the crate repository under
`.buildkite/hooks`.

Example of post-checkout hook that downloads and extracts a bzImage:

```bash
#!/bin/bash

DEB_NAME="linux-image-4.9.0-9-amd64_4.9.168-1_amd64.deb"
DEB_URL="http://ftp.debian.org/debian/pool/main/l/linux/${DEB_NAME}"

REPO_PATH="${BUILDKITE_BUILD_CHECKOUT_PATH}"
DEB_PATH="${REPO_PATH}/${DEB_NAME}"
EXTRACT_PATH="${REPO_PATH}/src/bzimage-archive"
BZIMAGE_PATH="${EXTRACT_PATH}/boot/vmlinuz-4.9.0-9-amd64"

mkdir -p ${EXTRACT_PATH}

> wget $DEB_URL -P ${REPO_PATH}

Reviewer: ```suggestion
wget ${DEB_URL} -P ${REPO_PATH}
```

dpkg-deb -x ${DEB_PATH} ${EXTRACT_PATH}

mv ${BZIMAGE_PATH} ${REPO_PATH}/src/bzimage
rm -r ${EXTRACT_PATH}
rm -f ${DEB_PATH}
```

In this example the post-checkout hook downloads a deb image, extracts the
contents of it and places it in `linux-loader/src/bzimage`. The unit tests
will use the relative path `src/bzimage` which does not depend on the image
being downloaded.

## Integration Tests

> Besides the one-liner tests defined in the

Reviewer: ```suggestion
In addition to the one-liner tests defined in the
```

# rust-vmm-ci

> The `rust-vmm-ci` contains [integration tests](#integration-tests) and

Reviewer: nit
```suggestion
The `rust-vmm-ci` repository contains [integration tests](#integration-tests) and
```

  separated by one coma. If the repository does not have any autogenerated
  files, *exclude_path* should be an empty string.
- *crate_features*: `cargo kcov` does not build crate features by default. To
  get the coverage report including optional features these need to be
  specified in *crate_features* separated by coma. If the crate does not have
  any features, this field should be empty.

This file is required for the coverage integration so it needs to be added
to the repository as well.

3. Create a new pipeline definition in Buildkite. For this step ask one of the
rust-vmm Buildkite [admins](CODEOWNERS) to create one for you. Add a pipeline
step that is uploading the rust-vmm-ci pipeline:

```bash
buildkite-agent pipeline upload rust-vmm-ci/.buildkite/pipeline.yml
```

4. The code owner of the repository will have to setup a WebHook for triggering
the CI on Pull Requests and pushes.

## Buildkite Pipeline

The [Buildkite](https://buildkite.com) pipeline is the definition of tests to
be run as part of the CI. It includes steps for running unit tests and linters
(including coding style checks), and computing the coverage.

> Currently the tests run on a Linux machine on x86_64 and aarch64.

Reviewer 2: Nit: `a` here feels like it's running on a specific machine. I'd rephrase it as:
`Currently the tests can run on Linux x86_64 and aarch64 hosts`
